File: src/main/java/com/laamella/code_state_machine/io/DotOutput.java
package com.laamella.code_state_machine.io;

import com.laamella.code_state_machine.StateMachine;
import com.laamella.code_state_machine.Transition;

/**
 * Creates a "dot" diagram of the state machine.
 */
public class DotOutput<T, E, P extends Comparable<P>> {
	public String getOutput(final StateMachine<T, E, P>.MetaInformation metaInformation) {
		final StringBuilder output = new StringBuilder();
		output.append("digraph finite_state_machine {\n");
		output.append("\trankdir=LR;\n");
		output.append("\tsize=\"8,5\"\n");
		output.append("\tnode [shape = doublecircle, style=solid];");
		for (final T startState : metaInformation.getStartStates()) {
			output.append(" " + startState);
		}
		output.append(";\n");
		output.append("\tnode [shape = circle, style=dotted];");
		for (final T startState : metaInformation.getEndStates()) {
			output.append(" " + startState);
		}
		output.append(";\n");
		output.append("\tnode [shape = circle, style=solid];\n");
		for (final T sourceState : metaInformation.getSourceStates()) {
			for (final Transition<T, E, P> transition : metaInformation.getTransitionsForSourceState(sourceState)) {
				output.append("\t" + sourceState + " -> " + transition.getDestinationState());

				output.append(" [ label = \"" + transition.getCondition() + "\" ]");

				output.append(";\n");
			}
		}
		output.append("}\n");
		return output.toString();
	}
}

File: scala/src/main/scala/com/laamella/code_state_machine/io/DotOutput.scala
package com.laamella.code_state_machine.io

import com.laamella.code_state_machine.StateMachine
import com.laamella.code_state_machine.Transition

/**
 * Creates a simple "dot" diagram of the state machine. Start states are double
 * circles, end states are dotted circles, entry and exit events are not shown.
 */
class DotOutput[State, Event, Priority <: Ordered[Priority]] {
	def getOutput(machine: StateMachine[State, Event, Priority] ): String = {

		val output = new StringBuilder()
		output.append("digraph finite_state_machine {\n")
		output.append("\trankdir=LR;\n")
		output.append("\tsize=\"8,5\"\n")
		if (machine.startStates.nonEmpty) {
			output.append("\tnode [shape = doublecircle, style=solid];")
			for (startState <- machine.startStates) {
				output.append(" " + startState)
			}
			output.append(";\n")
		}
		if (machine.endStates.nonEmpty) {
			output.append("\tnode [shape = circle, style=dotted];")
			for (startState <- machine.endStates) {
				output.append(" " + startState)
			}
			output.append(";\n")
		}
		output.append("\tnode [shape = circle, style=solid];\n")
		for (sourceState <- machine.transitions.keys) {
			for (transition <- machine.transitions(sourceState)) {
				output.append("\t" + sourceState + " -> " + transition.destinationState)
				output.append(" [ label = \"" + transition.conditions + "\" ]")
				output.append(";\n")
			}
		}
		output.append("}\n")
		output.toString()
	}
}
